### ISSUE TITLE
concurrency issue?

### ISSUE THREAD

**Issue by christopher-johnson** [trellis-ldp/trellis-io-jena#1]
I ran into [this issue](http://mail-archives.apache.org/mod_mbox/jena-users/201704.mbox/%3CBFA357FF-B33C-4FE3-A1EF-A8C244ADEF6C@email.virginia.edu%3E)  today using this function:
```java
ioService.read(stream, null, NTRIPLES).forEach(graph::add);
```
If I run the stream directly through RDFMgr, like this:
```java
RDFDataMgr.read(model, stream, null,  lang)
```
it works.   The graph stream is quite large and I am getting as JSON-LD over the wire.  I suspect that is related with the NamespaceService mapping [here](https://github.com/trellis-ldp/trellis-io-jena/blob/master/src/main/java/org/trellisldp/io/JenaIOService.java#L153-L158)

**Comment by acoburn** [comment]
@christopher-johnson are you reading multiple streams into the same `graph` object? If so, you'd need to coordinate those.

**Comment by christopher-johnson** [comment]
no, I do not think so.  The logic is nothing out of the ordinary:
```java
URL uri = new URL(testResource);
InputStream is = expandDocumentToN3(uri);
graph = getGraph(is);
```
the expand part uses `JsonLdProcessor.toRDF` and then:
```java
public static Graph getGraph(InputStream stream) {
    final Graph graph = rdf.createGraph();
    ioService.read(stream, null, NTRIPLES).forEach(graph::add);
    return graph;
}
```
 the problem always occurs somewhere in `read()`

when I change `getGraph() `to this:
```java
final Model model = createDefaultModel();
final Lang lang = rdf.asJenaLang(NTRIPLES).orElseThrow(() ->
 new RuntimeRepositoryException("Unsupported RDF Syntax: " + NTRIPLES.mediaType));
RDFDataMgr.read(model, stream, null,  lang);
return rdf.asGraph(model)
```
it works.  It only happens on large models.

**Comment by acoburn** [comment]
And one more thing to confirm before I dig into this: the `rdf` object you're using -- is that a `JenaRDF` object or a `SimpleRDF` object? `SimpleRDF` does a very poor job of handling concurrency, and it's really only good for simple sorts of tests.

**Comment by christopher-johnson** [comment]
the rdf object is a static final JenaRDF.  The `oh dear, already have a slot` exception is rather odd.  

One other possibility for a concurrency issue is that after I get the graph, I also pass it to this:
```java
org.apache.jena.graph.Graph jenaGraph = rdf.asJenaGraph(graph);
model = ModelFactory.createModelForGraph(jenaGraph);
````
though, in debugging, I noticed that getGraph() returned null...

**Comment by acoburn** [comment]
I am pretty certain that this is related to the namespace mapping service. I have updated the code there to use a `ConcurrentHashMap` internally and have cut a 0.2.1 release, which is now available on Maven Central. If that does not fix the concurrency error, please re-open this issue.

**Comment by christopher-johnson** [comment]
I have written a bulk ingest harvester and I have again experienced the "oh dear" exception, but this time directly in trellis-http.  I attach a stack trace for reference.  [stacktrace.txt](https://github.com/trellis-ldp/trellis-io-jena/files/1466178/stacktrace.txt)

I now believe that it has to do with the nondeterminism of `Stream.forEach`.  When used by `ContentBearingHandler` its parallelism could cause "overlapping writes" in `TrellisDataset.add` that lead to the "oh dear" exception.  When I change the [forEach](https://github.com/trellis-ldp/trellis-http/blob/master/src/main/java/org/trellisldp/http/impl/ContentBearingHandler.java#L84) to `forEachOrdered`, I have not experienced the issue.  

Not sure about the performance negatives to not using parallelism for this function, but with a dataset ingest operation, there is definitely a sequence.

**Comment by acoburn** [comment]
Yes, that should definitely be `forEachOrdered`. I'll fix this and also look through the rest of the code for similar cases.

**Comment by acoburn** [comment]
See https://github.com/trellis-ldp/trellis-http/commit/f3a42ce2e9ea83740d7fed7172deb55b4311830e and https://github.com/trellis-ldp/trellis-http/commit/bbc0db4285dc224ba11d3f48af5a019b6b0c2b7a

**Comment by acoburn** [comment]
@christopher-johnson @ajs6f I've added some new methods in the `trellis-api` module to make it easier to collect Triples/Quads into Graph/Dataset objects: https://github.com/trellis-ldp/trellis-api/blob/master/src/main/java/org/trellisldp/api/RDFUtils.java#L57-L81

I haven't started using this code in the downstream modules, but let me know if you have any comments on it. In most cases, this should be better than using `forEachOrdered`.

**Comment by acoburn** [comment]
One question I have is whether such methods should become part of the `commons-rdf` project.

**Comment by ajs6f** [comment]
I would have used `right.iterate().forEach(left::add);`, but as for Commons, I'm not sure. What's the advantage over 
```
Graph graph = ...
Stream<Triple> trips = ...
trips.forEach(graph::add);
```
?

**Comment by acoburn** [comment]
@ajs6f because if the stream is parallel, it can trigger a concurrency error.

**Comment by ajs6f** [comment]
Yeah, I don't see where the concurrency issue got fixed? It seems like this kind of throws the baby out with the bathwater?

**Comment by acoburn** [comment]
Basically, with this code:

```java
Graph graph = ...
Stream<Triple> trips = ...
trips.forEach(graph::add);
```
we get a concurrency error if the stream is parallel (given a sufficiently large stream).

One fix is to turn the stream into a sequential stream: `trips.sequential().forEach(graph::add)` or `trips.forEachOrdered(graph::add)`, which undermines the possibility of parallelism in the processing pipeline.

using a `Collector` (as I understand `Collector`s) the stream can still be parallel, but there is now a defined mechanism to fold that parallel stream into a single mutable reduction, while maintaining some level of coordination on that target object.

**Comment by ajs6f** [comment]
Right, and I would make the 'Graph' threadsafe instead. (By wrapping it.)

**Comment by ajs6f** [comment]
Or with an improvement to the Jena 'Graph' impl.

**Comment by ajs6f** [comment]
TBC, I am in no way opposed to Collectors-- https://issues.apache.org/jira/plugins/servlet/mobile#issue/JENA-1391

I just don't think they get to the root of the problem here.

**Comment by christopher-johnson** [comment]
Great work Aaron.  I did an quick performance analysis using your test to compare sequential() and parallel() and have attached the result output for reference.  I set the limit to 500k triples. parallel() took 9207ms and sequential() 5881ms.  I implemented a counter to see how parallel() nondeterminism works, and it is interesting to see when the count stops being sequential.  
I have a Intel(R) Core(TM) i5-4310M CPU @ 2.70GHz.
[output-parallel.txt](https://github.com/trellis-ldp/trellis-io-jena/files/1484812/output-parallel.txt)
[output-sequential.txt](https://github.com/trellis-ldp/trellis-io-jena/files/1484813/output-sequential.txt)

I wonder about the benefit of parallelism for Graph since the work to create a triple is relatively small.  Up to you :+1:
